libbeat/outputs/shipper/shipper.go

```diff
@@ -164,90 +193,135 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
+			Uuid:   s.serverID,
+			Events: toSend,
+		})
 
-	if indexReply.GetUuid() != c.serverID {
-		batch.Cancelled()
-		st.Cancelled(len(events))
-		return fmt.Errorf("acknowledgement failed due to a connection to a different server %s, expected %s", indexReply.Uuid, c.serverID)
-	}
+		if status.Code(err) != codes.OK {
+			batch.Cancelled()         // does not decrease the TTL
+			st.Cancelled(len(events)) // we cancel the whole batch not just non-dropped events
+			return fmt.Errorf("failed to publish the batch to the shipper, none of the %d events were accepted: %w", len(convertedEvents), err)
+		}
 
-	for indexReply.PersistedIndex < publishReply.AcceptedIndex {
-		indexReply, err = ackClient.Recv()
-		if err != nil {
-			return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+		// with a correct server implementation should never happen, this error is not recoverable
+		if int(publishReply.AcceptedCount) > len(nonDroppedEvents) {
+			return fmt.Errorf(
+				"server returned unexpected results, expected maximum accepted items %d, got %d",
+				len(nonDroppedEvents),
+				publishReply.AcceptedCount,
+			)
 		}
+		toSend = toSend[publishReply.AcceptedCount:]
+		lastAcceptedIndex = publishReply.AcceptedIndex
 	}
 
-	retries := nonDroppedEvents[publishReply.AcceptedCount:]
-	if len(retries) == 0 {
-		batch.ACK()
-		st.Acked(len(nonDroppedEvents))
-		c.log.Debugf("%d events have been accepted, %d dropped", len(nonDroppedEvents), droppedCount)
-	} else {
-		batch.RetryEvents(retries) // decreases TTL unless guaranteed delivery
-		st.Failed(len(retries))
-		c.log.Debugf("%d events have been accepted, %d sent for retry, %d dropped", publishReply.AcceptedCount, len(retries), droppedCount)
-	}
+	s.log.Debugf("%d events have been accepted, %d dropped", len(nonDroppedEvents), droppedCount)
+
+	s.pendingMutex.Lock()
+	s.pending = append(s.pending, pendingBatch{
+		batch:        batch,
+		index:        lastAcceptedIndex,
+		serverID:     s.serverID,
+		droppedCount: droppedCount,
+	})
+	s.pendingMutex.Unlock()
 
 	return nil
 }
 
 // Close closes the connection to the shipper server.
 // Also, implements `outputs.Client`
-func (c *shipper) Close() error {
-	if c.client == nil {
+func (s *shipper) Close() error {
+	if s.client == nil {
 		return fmt.Errorf("connection is not established")
 	}
-	err := c.conn.Close()
-	c.conn = nil
-	c.client = nil
+	s.backgroundCancel()
+	err := s.conn.Close()
+	s.conn = nil
+	s.client = nil
 
 	return err
 }
 
 // String implements `outputs.Client`
-func (c *shipper) String() string {
+func (s *shipper) String() string {
 	return "shipper"
 }
 
+func (s *shipper) ackLoop(ctx context.Context) error {
+	st := s.observer
+
+	// so we explicitly close the stream when exit the function
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s.clientMutex.Lock()
+	ackClient, err := s.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
```

Comment:
There should probably be a nil check on `s.client` (it will _almost_ never be nil but it's technically possible)

Reply:
Well it won't be possible after I make the change you suggested in https://github.com/elastic/beats/pull/32785/files#r954279971